Add support for use_inline_powershell for powershell_script

jaymzjulian opened the pull request:
## Description

This adds support for the powershell_script resource to use powershell.dll directly instead of going through the powershell_out process - this is _significantly_ faster if you're doing a lot of small powershell changes.  For our environment, doing things this way instead of via powershell_out yields a 28% reducition in chef run time.

It is not, and should not be, default behaviour because there are a bunch of semantic differences - most notably, you can't stream output, and you can't switch user.  But for a significant number of powershell use cases, the large performance improvement makes this worth the compromise.  This is particularly true for idempotent runs, where executing guards can be a significant user of cycles from constantly starting powershell.exe/pwsh.exe

It can be seen to work with the following chef-apply script:

```
powershell_script 'guardme - should not run' do
  code <<-EOH
Write-Output "hi"
EOH
  use_inline_powershell true
  only_if '$False'
end

powershell_script 'guardme2 - should run' do
  code <<-EOH
Write-Output "hi"
EOH
  use_inline_powershell true
  only_if '$True'
end

powershell_script 'guardme3 - should not run' do
  code <<-EOH
Write-Output "hi"
EOH
  use_inline_powershell true
  only_if 'throw_exception'
end

powershell_script 'using creates' do
  code <<-EOH
Write-Output "hi"
EOH
  use_inline_powershell true
  creates "c:/windows"
end

powershell_script "fail on pourpose" do
  code "return 1"
  ignore_failure true
end
```

which outputs
```
Recipe: (chef-apply cookbook)::(chef-apply recipe)
  * powershell_script[guardme - should not run] action run (skipped due to only_if)
  * powershell_script[guardme2 - should run] action run
    - execute direct powershell guardme2 - should run
  * powershell_script[guardme3 - should not run] action run (skipped due to only_if)
  * powershell_script[using creates] action run (up to date)
  * powershell_script[fail on pourpose] action run

    ================================================================================
    Error executing action `run` on resource 'powershell_script[fail on pourpose]'
    ================================================================================
```


## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Chore (non-breaking change that does not add functionality or fix an issue)

## Checklist:
- [x] I have read the **CONTRIBUTING** document.
- [x] I have run the pre-merge tests locally and they pass.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [ ] If `Gemfile.lock` has changed, I have used `--conservative` to do it and included the full output in the Description above.
- [x] All new and existing tests passed.
- [x] All commits have been signed-off for [the Developer Certificate of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

tpowell-progress commented on lib/chef/provider/powershell_script.rb:
```diff
@@ -27,8 +27,71 @@ class PowershellScript < Chef::Provider::WindowsScript
+            stderr = nil
+            Timeout.timeout(new_resource.timeout) do
+              r = powershell_exec!(wrapper_script, new_resource.interpreter.to_sym)
+              # Split out the stdout/returncode format if needed
```
I think just break "returncode" into two words here

tpowell-progress commented on lib/chef/provider/powershell_script.rb:
```diff
@@ -27,8 +27,71 @@ class PowershellScript < Chef::Provider::WindowsScript
+              # Split out the stdout/returncode format if needed
+              return_code = r.result
+              if return_code.is_a?(Array)
+                # Why to_s?  Becuase if the only powershell output is an object,
```
typo

tpowell-progress commented on lib/chef/provider/powershell_script.rb:
```diff
@@ -130,7 +193,78 @@ def validate_script_syntax!
       # (success, where we return 0) or false (where we return 1).
       def wrapper_script
-        <<~EOH
+        # The script looks very slightly different with inline powershell since
```
Can we extract these two options into their own methods? Maybe something like `inline_powershell_wrapper...` and `shell_out_wrapper...`? The method is a bit long to start with, so the if/else is harder to read due to the length.

jaymzjulian replied:
absolutely can

tpowell-progress commented on lib/chef/resource/powershell_script.rb:
```diff
@@ -21,7 +21,7 @@ class Chef
     class PowershellScript < Chef::Resource::WindowsScript
 
-      set_guard_inherited_attributes(:interpreter)
+      set_guard_inherited_attributes(:interpreter, :use_inline_powershell)
```
I think this may be contributed to the Linux test failures in buildkite

jaymzjulian replied:
yeah, it seems so - I'm not sure how I missed that one, since it even tripped locally for me this time... I'm guessing i was dumb and ran the test before the change!

jaymzjulian replied:
looks like updating the test that consumes this fixed this.  woo!  thanks for the heads up!

tpowell-progress commented on lib/chef/provider/powershell_script.rb:
```diff
@@ -27,8 +27,71 @@ class PowershellScript < Chef::Provider::WindowsScript
-        validate_script_syntax!
-        super()
+        Chef::Log.debug("using inline impl: #{new_resource.use_inline_powershell}")
+        if new_resource.use_inline_powershell
```
Now that I can see the forest, I will more thoughts. Stay tuned.

tpowell-progress replied:
Ok, so I think the general flow and logic is good, but I really would like to see this chunked up and modularized better.

jaymzjulian replied:
> Ok, so I think the general flow and logic is good, but I really would like to see this chunked up and modularized better.

I can definitely look at doing that - there are a couple of places I can see to do that:
a) moving the inline powershell version into a function, rather than in action :run, but 
b) modularizing the stuff i cut/pasted from execute - I avoided that because I didn't think it was a good idea to mess with execute.rb generally, BUT honestly, looking at it with fresh eyes, that seems like a sane thing to do?

tpowell-progress replied:
@jaymzjulian ok, the cut/paste modularization might be better served for a *second* PR, maybe even after the next Chef release. @dafyddcrosby What do you think about introducing this for the next release and then refactoring as a second pass?

tpowell-progress reviewed (changes requested):
@jaymzjulian Still haven't digested everything, but one request I can already think of: Can we have a `spec/functional/resource/powershell_script_spec.rb` test that validates inputs and outputs of inline vs. shell out remain identical between the changes (non-functional aspects such as performance are an acknowledged difference).

tpowell-progress commented on lib/chef/provider/powershell_script.rb:
```diff
@@ -129,7 +192,80 @@ def validate_script_syntax!
       # executed, otherwise 0 or 1 based on whether $? is set to true
       # (success, where we return 0) or false (where we return 1).
-      def wrapper_script
+      def inline_powershell_wrapper_script
```
Thinking this can potentially be deduplicated by some sort of templating or builder logic so that the behavior stays consistent between them.  cc: @dafyddcrosby

jaymzjulian replied:
yeah, i think that's sane to do - i'll make that happen

jaymzjulian replied:
In the latest push, this is simplified to a single function with the two changes as simple conditions at those line points, which seemed much easier to read to my eye.

johnmccrae commented:
@jaymzjulian I have a naïve question - what does this do that powershell_exec, as provided by the chef-powershell gem, doesn't do?

tpowell-progress commented on lib/chef/provider/powershell_script.rb:
```diff
@@ -27,8 +27,71 @@ class PowershellScript < Chef::Provider::WindowsScript
+            Timeout.timeout(new_resource.timeout) do
+              r = powershell_exec!(inline_powershell_wrapper_script, new_resource.interpreter.to_sym)
+              # Split out the stdout/return code format if needed
+              return_code = r.result
```
Seems like a lot of this code is already encapsulated by `chef/chef-powershell-shim:chef-powershell/lib/chef-powershell/powershell.rb` and friends... the timeout, the capturing of errors, etc.. (see [exec](https://github.com/chef/chef-powershell-shim/blob/main/chef-powershell/lib/chef-powershell/powershell.rb#L128))

jaymzjulian replied:
The honest answer to that I knew about powershell_exec, which doens't, but not Powershell.exec, which does the timeout and some of the demunging stuff.  I could absolutely move to that implentation, and use the timeout stuff there

jaymzjulian replied:
I'm fixing up my tests before pushing more code at this (working out how to make rspec acutally succesfully call FFI stuff... whee!), but I acutally do have a useful answer to this now - powershell_exec actually calls the stuff in chef-powershell, but also has the ability to switch interpreter on its own.  I _have_ refactored it to use the inbuilt timeout stuff, however, and am refactoring out some of the error handling complexirty

jaymzjulian replied:
Oh, to answer the actual question, the thing this does that just calling powershell_exec doesn't is be the powershell_script resource - since powershell_exec isn't a resource, to use it in an expected converge-time way, users either have to wrap it in ruby_block, have it in a unified mode resource, or wrap it in some other resource.  The point of this is to have a simple property that can be set to do things the performant but not-quite-as-featured way, rather than having all of that opacity.

This is particularly true if you want simple guards - you end up with messes like:

```
fast_powershell_thing "thing" do
   code "thing"
   not_if { powershell_exec!("guardthing") }
end
```

or similar monstrosities.  My actual initial implementation involved literally overriding only_if and not_if in the resource to make this read sanely, which does work in a custom resource, but that's a terrible, terrible pattern that I really want away from.

tpowell-progress replied:
> I'm fixing up my tests before pushing more code at this (working out how to make rspec acutally succesfully call FFI stuff... whee!), but I acutally do have a useful answer to this now - powershell_exec actually calls the stuff in chef-powershell, but also has the ability to switch interpreter on its own. I _have_ refactored it to use the inbuilt timeout stuff, however, and am refactoring out some of the error handling complexirty

@jaymzjulian are your test updates ready for review?

jaymzjulian replied:
> @jaymzjulian are your test updates ready for review?

they are

jaymzjulian commented:
Oh course this breaks all tests on non windows:(. Will fix in the morning

jaymzjulian commented:
> Oh course this breaks all tests on non windows:(. Will fix in the morning

tests looking better now - I don't think it was actually my code that broke them.  For good measure I also added some kitchen tests to exercise this as well as the unit tests, which also pass.  huzzah and all that.

tpowell-progress commented:
@jaymzjulian can you rebase on main just to clear up the noise from the test breakage from the point in main from which you branched?

jaymzjulian commented:
> > > @jaymzjulian can you rebase on main just to clear up the noise from the test breakage from the point in main from which you branched?
> > 
> > 
> > oh, sorry, i didn't realise the tests were failing because of my bad rebase - i've rebased to main now... fingers crossed
> 
> Looks like the [Windows Unit Tests](https://buildkite.com/chef-oss/chef-chef-main-verify/builds/16148#018bf416-ab34-4d7f-ab35-936fe96f213e/522-17535) are failing on the changes... looks like existing tests on first glance which makes me think the change is bleeding outside of the setting for inline.

Ahh, I see what I did there.... I protected the function `code` during a re-org.  Looking at the diff, that seems to be the only leaking change - I did make the tests resiliant to that as well, since I think that probably _should_ be protected, but I also reverted that to be unprotected as it was before

tpowell-progress commented:
> > > > @jaymzjulian can you rebase on main just to clear up the noise from the test breakage from the point in main from which you branched?
> > > 
> > > 
> > > oh, sorry, i didn't realise the tests were failing because of my bad rebase - i've rebased to main now... fingers crossed
> > 
> > 
> > Looks like the [Windows Unit Tests](https://buildkite.com/chef-oss/chef-chef-main-verify/builds/16148#018bf416-ab34-4d7f-ab35-936fe96f213e/522-17535) are failing on the changes... looks like existing tests on first glance which makes me think the change is bleeding outside of the setting for inline.
> 
> Ahh, I see what I did there.... I protected the function `code` during a re-org. Looking at the diff, that seems to be the only leaking change - I did make the tests resiliant to that as well, since I think that probably _should_ be protected, but I also reverted that to be unprotected as it was before

@jaymzjulian everything looks good so far. will wait for the Community PR review since we're already there [again]
